internal/shim-sev/src/hostmap.rs

```diff
@@ -0,0 +1,241 @@
+                                    return;
+                                }
+
+                                let page: *mut HostMemListPage = page_res.unwrap().as_ptr() as _;
+
+                                page.write_bytes(0, 1);
+                                page
+                            };
+
+                            (*last_page).header.next = Some(&mut *new_page);
+                            this.num_pages = this.num_pages.checked_add(1).unwrap();
+                            new_page
+                        }
+                        Some(ref mut p) => *p as *mut _,
+                    };
+                }
+            }
+        }
+
+        // While updating the HostMap the syscall proxying can't be used,
+        // so disable any debug printing possibly happening while allocating
+        // memory in `inner_extend_slots`.
+        unsafe {
+            SHIM_CAN_PRINT.store(false, Ordering::Release);
+        }
+
+        inner_extend_slots(self.write(), mem_slots, allocator);
+
+        unsafe {
+            SHIM_CAN_PRINT.store(true, Ordering::Release);
+        }
+    }
+
+    /// Translate a shim physical unencrypted address to a host virtual address
+    pub fn shim_phys_to_host_virt<U>(
+        &self,
+        shim_phys: ShimPhysUnencryptedAddr<U>,
+    ) -> HostVirtAddr<U> {
+        let this = self.read();
+        let addr: usize = shim_phys.raw().raw() as _;
+
+        let virt_addr = this.get_virt_addr(addr).unwrap_or_else(|| {
+            panic!(
+                "Trying to get virtual offset from unmmapped location {:#?}",
+                addr
+            )
+        });
+
+        unsafe { HostVirtAddr::new(Address::<u64, U>::unchecked(virt_addr as _)) }
+    }
+
+    /// set the initial map entry
+    pub fn first_entry(&self, start: usize, host_span: Span<usize>) {
+        let mut this = self.write();
+        this.host_mem.ent[0].shim.start = start;
+        this.host_mem.ent[0].shim.end = start.checked_add(host_span.count).unwrap();
+        this.host_mem.ent[0].virt_start = host_span.start;
+    }
+
+    /// Add a new map entry
+    pub fn new_entry(&self, host_span: Span<usize>) -> Option<Line<usize>> {
+        let mut this = self.write();
+        let mut free = &mut this.host_mem;
+        let mut last_end: usize = 0;
+
+        loop {
+            for i in free.ent.iter_mut() {
+                if i.shim.end == 0 {
+                    i.virt_start = host_span.start;
+                    i.shim.start = last_end;
+                    i.shim.end = i.shim.start.checked_add(host_span.count).unwrap();
+                    return Some(i.shim);
+                }
+                last_end = i.shim.end;
+            }
+
+            // we have reached the end of the free slot page
+            // advance to the next page
+            if let Some(f) = free.header.next.as_deref_mut() {
+                free = f;
+            } else {
+                return None;
+            }
```

Comment:
This can be a constant-time operation if the hostmap carries a "next available index" field. In order to avoid gaps when removing a mapping it'll be necessary to replace the "to-be-deleted" element with the final element in the list and then update the "next available index" field accordingly.

Reply:
Hmm, ballooning doesn't happen that often and I think it is time critical here. Also, I don't think we will support memory shrinking.

Happy to review and accept PRs, though :)